File: generated_code/py/program_3_2.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Dictionary to store the last seen index of each character
        char_map = {}
        max_length = 0
        start = 0  # Left pointer of the sliding window
        
        # End pointer iterates through the string
        for end, char in enumerate(s):
            # If the character is already in the map and within the current window,
            # slide the start pointer to the right of its last occurrence
            if char in char_map and char_map[char] >= start:
                start = char_map[char] + 1
            
            # Update the last seen position of the character
            char_map[char] = end
            
            # Update the maximum length found so far
            # Window length is calculated as (end - start + 1)
            max_length = max(max_length, end - start + 1)
            
        return max_length

File: generated_code/cpp/program_3_2.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // We use an array to store the last seen position of each character.
        // Since the input consists of standard ASCII characters (letters, digits, symbols, spaces),
        // a size of 128 is sufficient. Initialize all positions to -1.
        int last_seen[128];
        std::fill_n(last_seen, 128, -1);
        
        int max_len = 0;
        int start = 0; // Left boundary of the sliding window
        
        // Expand the right boundary of the window
        for (int end = 0; end < s.length(); ++end) {
            char current_char = s[end];
            
            // If the character was seen inside the current window, 
            // shrink the window by moving 'start' past its previous occurrence.
            if (last_seen[current_char] >= start) {
                start = last_seen[current_char] + 1;
            }
            
            // Update the last seen position of the character
            last_seen[current_char] = end;
            
            // Calculate and track the maximum window size achieved
            max_len = std::max(max_len, end - start + 1);
        }
        
        return max_len;
    }
};